File: apps/api/app/schemas/report.py
import datetime as dt
import uuid
from decimal import Decimal

from pydantic import BaseModel

from app.schemas.dashboard import ExpenseByCategory


class MonthlyReport(BaseModel):
    year: int
    month: int
    total_income: Decimal
    total_expenses: Decimal
    net_balance: Decimal
    expenses_by_category: list[ExpenseByCategory]
    daily_spending: list[dict]


class YearlyReportMonth(BaseModel):
    month: int
    income: Decimal
    expenses: Decimal


class YearlyReport(BaseModel):
    year: int
    total_income: Decimal
    total_expenses: Decimal
    months: list[YearlyReportMonth]
    expenses_by_category: list[ExpenseByCategory]


class CategoryHistoryPoint(BaseModel):
    period: str
    amount: Decimal


class CategoryReport(BaseModel):
    category_id: uuid.UUID
    category_name: str
    history: list[CategoryHistoryPoint]


class AIAnalysisJobResponse(BaseModel):
    job_id: str


class AIAnalysisStatusResponse(BaseModel):
    status: str
    result: str | None = None


class AIAnalysisRequest(BaseModel):
    period_start: dt.date
    period_end: dt.date

File: apps/web/src/types/report.ts
import type { ExpenseByCategory } from './dashboard'

export interface MonthlyReport {
  year: number
  month: number
  total_income: number
  total_expenses: number
  net_balance: number
  expenses_by_category: ExpenseByCategory[]
  daily_spending: { date: string; expenses: number; income: number }[]
}

export interface YearlyReportMonth {
  month: number
  income: number
  expenses: number
}

export interface YearlyReport {
  year: number
  total_income: number
  total_expenses: number
  months: YearlyReportMonth[]
  expenses_by_category: ExpenseByCategory[]
}

export interface AIAnalysisJob {
  job_id: string
}

export interface AIAnalysisStatus {
  status: 'PENDING' | 'STARTED' | 'SUCCESS' | 'FAILURE'
  result: string | null
}
